contrib/jms/src/main/java/com/google/cloud/dataflow/contrib/jms/JmsIO.java

```diff
@@ -0,0 +1,532 @@
+                return new Bound<T>(name, connectionFactory, queue, topic, coder);
+            }
+
+            public <X> Bound<X> withCoder(Coder<X> coder) {
+                return new Bound<>(name, connectionFactory, queue, topic, coder);
+            }
+
+            @Override
+            public PDone apply(PCollection<T> input) {
+                if (connectionFactory == null) {
+                    throw new java.lang.IllegalStateException("need to set the JMS " +
+                            "connection factory of a " +
+                            "JmsIO.Write transform");
+                }
+                if (queue == null && topic == null) {
+                    throw new java.lang.IllegalStateException("need to set the JMS " +
+                            "destination (queue or topic) of " +
+                            "a JmsIO.Write transform");
+                }
+                input.apply(ParDo.of(new JmsWriter()));
+                return PDone.in(input.getPipeline());
+            }
+
+            @Override
+            protected Coder<T> getDefaultOutputCoder() {
+                return coder;
+            }
+
+            public ConnectionFactory getConnectionFactory() {
+                return connectionFactory;
+            }
+
+            public String getQueue() {
+                return queue;
+            }
+
+            public String getTopic() {
+                return topic;
+            }
+
+            public Coder<T> getCoder() {
+                return coder;
+            }
+
+            private class JmsWriter extends DoFn<T, Void> {
+                private transient Connection connection;
+                private transient Session session;
+
+                @Override
+                public void startBundle(Context c) throws Exception {
+                    connection = connectionFactory.createConnection();
+                    session = connection.createSession(false, Session.AUTO_ACKNOWLEDGE);
+                }
+
+                @Override
+                public void processElement(ProcessContext c) throws Exception {
+                    TextMessage message = session.createTextMessage(
+                            (String) CoderUtils.decodeFromByteArray(
+                                    getCoder(),
+                                    ((String) c.element()).getBytes()));
+                    MessageProducer producer;
+                    if (queue != null) {
+                        producer = session.createProducer(session.createQueue(queue));
+                    } else {
+                        producer = session.createProducer(session.createTopic(topic));
+                    }
+                    producer.send(message);
+                    producer.close();
+                }
+
+                @Override
+                public void finishBundle(Context c) throws Exception {
+                    if (session != null) {
+                        session.close();
```

Comment:
Is it important to make sure that `connection` is also closed even if `session.close()` throws an exception?